drivers/aip31068/aip31068.c

```diff
@@ -0,0 +1,516 @@
+    }
+
+    return _command(dev, CMD_ENTRY_MODE_SET | dev->_curr_entry_mode_set);
+}
+
+int aip31068_set_cursor_blinking_enabled(aip31068_t *dev, bool enabled)
+{
+    if (enabled) {
+        dev->_curr_display_control |= (1 << BIT_DISPLAY_CONTROL_CURSOR_BLINKING);
+    }
+    else {
+        dev->_curr_display_control &= ~(1 << BIT_DISPLAY_CONTROL_CURSOR_BLINKING);
+    }
+
+    return _command(dev, CMD_DISPLAY_CONTROL | dev->_curr_display_control);
+}
+
+int aip31068_set_cursor_visible(aip31068_t *dev, bool visible)
+{
+    if (visible) {
+        dev->_curr_display_control |= (1 << BIT_DISPLAY_CONTROL_CURSOR);
+    }
+    else {
+        dev->_curr_display_control &= ~(1 << BIT_DISPLAY_CONTROL_CURSOR);
+    }
+
+    return _command(dev, CMD_DISPLAY_CONTROL | dev->_curr_display_control);
+}
+
+int aip31068_set_cursor_position(aip31068_t *dev, uint8_t row, uint8_t col)
+{
+    col = ((row == 0) ? (col | 0x80) : (col | 0xc0));
+    uint8_t data[] = { 0x80, col };
+
+    return _device_write(dev, data, 2);
+}
+
+int aip31068_set_text_insertion_mode(aip31068_t *dev,
+                                     aip31068_text_insertion_mode_t mode)
+{
+    if (mode == RIGHT_TO_LEFT) {
+        dev->_curr_entry_mode_set &= ~(1 << BIT_ENTRY_MODE_INCREMENT);
+    }
+    else {
+        dev->_curr_entry_mode_set |= (1 << BIT_ENTRY_MODE_INCREMENT);
+    }
+
+    return _command(dev, CMD_ENTRY_MODE_SET | dev->_curr_entry_mode_set);
+}
+
+int aip31068_move_cursor_left(aip31068_t *dev)
+{
+    uint8_t cmd = CMD_CURSOR_DISPLAY_SHIFT;
+    cmd &= ~(1 << BIT_CURSOR_DISPLAY_SHIFT_DIRECTION);
+
+    return _command(dev, cmd);
+}
+
+int aip31068_move_cursor_right(aip31068_t *dev)
+{
+    uint8_t cmd = CMD_CURSOR_DISPLAY_SHIFT;
+    cmd |= (1 << BIT_CURSOR_DISPLAY_SHIFT_DIRECTION);
+
+    return _command(dev, cmd);
+}
+
+int aip31068_scroll_display_left(aip31068_t *dev)
+{
+    uint8_t cmd = CMD_CURSOR_DISPLAY_SHIFT;
+    cmd |= (1 << BIT_CURSOR_DISPLAY_SHIFT_SELECTION);
+    cmd &= ~(1 << BIT_CURSOR_DISPLAY_SHIFT_DIRECTION);
+
+    return _command(dev, cmd);
+}
+
+int aip31068_scroll_display_right(aip31068_t *dev)
+{
+    uint8_t cmd = CMD_CURSOR_DISPLAY_SHIFT;
+    cmd |= (1 << BIT_CURSOR_DISPLAY_SHIFT_SELECTION);
+    cmd |= (1 << BIT_CURSOR_DISPLAY_SHIFT_DIRECTION);
+
+    return _command(dev, cmd);
+}
+
+int aip31068_set_custom_symbol(aip31068_t *dev,
+                               aip31068_custom_symbol_t custom_symbol,
+                               uint8_t charmap[])
+{
+    uint8_t location = custom_symbol;
+    int rc = _command(dev, CMD_SET_CGRAM_ADDR | (location << 3));
+
+    if (rc < 0) {
+        return rc;
+    }
+
+    int row_count = dev->params.font_size == FONT_SIZE_5x8 ? 8 : 10;
+
+    for (int i = 0; i < row_count; i++) {
+        rc = _data(dev, charmap[i]);
+
+        if (rc < 0) {
+            return rc;
+        }
+    }
+
+    // todo: return to old cursor position
+    return _command(dev, CMD_SET_DDRAM_ADDR);
```

Comment:
?

Reply:
Well the problem here is that the cursor returns to (0,0), which is problematic when executing `aip31068_set_custom_symbol` after text has been written on the display and the user expects the cursor to stay at the position. So we need either a note in the documentation or I try to get the cursor position. That's something I need to investigate first, because the datasheet only mentions write operations.

Reply:
Is it expected to change the cursor if you are changing the character? (Haven't read the datashit in detail), otherwise can you get the current cursor position?

Reply:
When calling `_command(dev, CMD_SET_DDRAM_ADDR);` to switch back to DDRAM then it will jump back to the base address. It seems like reading operations are not supported, so I can't read the cursor position. Keeping track of the cursor position manually would lead to a high code overhead which also would be quite error prone. I'd simply add a note in the documentation to call this function before printing any characters.

Reply:
ok!